src/markov.rs

```diff
@@ -30,99 +31,108 @@ impl Key for Markov {
 impl Markov {
     pub fn new() -> Markov {
         Markov {
-            assocs: HashMap::new(),
-            start: HashMap::new(),
+            word_map: HashMap::new(),
+            starting_words: HashMap::new(),
         }
     }
 
     pub fn parse(&mut self, string: &str) {
-        // TODO: sanitize string of stuff like URLs
+        let url_regex = Regex::new(r"[-a-zA-Z0-9@:%._\+~#=]{2,256}\.[a-z]{2,6}\b([-a-zA-Z0-9@:%_\+.~#?&//=]*)").unwrap();
+        let string = url_regex.replace_all(string, "");
+        
         let mut words = string.split(' ');
-
-        // Get first word
-        let mut prev: String;
+        let mut previous_word: String;
+        
         match words.next() {
-            Some(s) => {
-                prev = String::from(s);
-                let count = self.start.entry(prev.clone()).or_insert(0);
+            Some(word) => {
+                previous_word = String::from(word);
+                let count = self.starting_words.entry(previous_word.clone()).or_insert(0);
                 *count += 1;
             }
+
             None => {
-                // Message is empty
+                println!("Empty message");
                 return;
             }
         }
 
-        // Get second word
-        let mut next: String;
+        let mut next_word: String;
         match words.next() {
-            Some(s) => {
-                next = String::from(s);
-                self.associate(prev.clone(), next.clone());
+            Some(word) => {
+                next_word = String::from(word);
+                self.associate(previous_word.clone(), next_word.clone());
             }
+
             None => {
-                self.associate(prev.clone(), String::from(""));
+                self.associate(previous_word.clone(), String::from(""));
                 return;
             }
         }
 
         for word in words {
-            // TODO strip punctuation and stuff
-            prev = next;
-            next = String::from(word);
-            self.associate(prev.clone(), next.clone());
+            previous_word = next_word.clone();
+            next_word = String::from(word);
+            self.associate(previous_word, next_word.clone());
         }
-        self.associate(next, String::from(""));
+        self.associate(next_word.clone(), String::from(""));
     }
 
     #[inline]
-    fn associate(&mut self, prev: String, next: String) {
-        let probs = self.assocs.entry(prev).or_insert_with(HashMap::new);
-        let count = probs.entry(next).or_insert(0);
+    fn associate(&mut self, previous_word: String, next_word: String) {
+        let probability = self.word_map.entry(previous_word).or_insert_with(HashMap::new);
+        let count = probability.entry(next_word).or_insert(0);
         *count += 1;
     }
 
-    pub fn generate(&self, length: u32) -> Option<String> {
+    pub fn generate(&self, length: u32, starting_word: Option<&String>) -> Option<String> {
```

Comment:
This should be a different function imo. Probably something like `generate_from_start()`
Then have `generate()` choose the first word, and then call `generate_from_start()` as to not duplicate code.